<div>
<img src=https://www.institutedata.com/wp-content/uploads/2019/10/iod_h_tp_primary_c.svg width="300">
</div>

# Lab 4.1.2 Linear Regression II

## Prediction of House Price Using Linear Regression

### Data

The Sacramento real estate transactions file is a list of 985 real estate transactions in the Sacramento area reported over a five-day period, as reported by the Sacramento Bee.

In [2]:
import pandas as pd
import numpy as np

%matplotlib inline

### 1. Explore Data

Perform EDA on this dataset. Find Dependent and Independent Variables to predict house price.

In [57]:
# Read CSV
house_csv = "house_new.csv"
house = pd.read_csv(house_csv)
house.head()

street        city    zip state  beds  baths  sq__ft  \
0      3526 HIGH ST  SACRAMENTO  95838    CA     2      1     836   
1       51 OMAHA CT  SACRAMENTO  95823    CA     3      1    1167   
2    2796 BRANCH ST  SACRAMENTO  95815    CA     2      1     796   
3  2805 JANETTE WAY  SACRAMENTO  95815    CA     2      1     852   
4   6001 MCMAHON DR  SACRAMENTO  95824    CA     2      1     797   

          type                     sale_date  price   latitude   longitude  \
0  Residential  Wed May 21 00:00:00 EDT 2008  59222  38.631913 -121.434879   
1  Residential  Wed May 21 00:00:00 EDT 2008  68212  38.478902 -121.431028   
2  Residential  Wed May 21 00:00:00 EDT 2008  68880  38.618305 -121.443839   
3  Residential  Wed May 21 00:00:00 EDT 2008  69307  38.616835 -121.439146   
4  Residential  Wed May 21 00:00:00 EDT 2008  81900  38.519470 -121.435768   

   actual_minus_average_price  predicted_house_prices      residuals  
0               170962.965347           135519.056164  -76297.056164  
1               161972.965347           176838.917832 -108626.917832  
2               161304.965347           130525.719406  -61645.719406  
3               160877.965347           137516.390867  -68209.390867  
4               148284.965347           130650.552825  -48750.552825

### 2. Predict Price

We are going to predict `price` from available information.

#### 2.1 What is the Target Variable? Why?

In [31]:
# ANSWER

#### 2.2 List all possible variables which might be Independent/Predictor variable.

In [32]:
# ANSWER

#### 2.3 Find correlation between variables

Find correlation between variables. Which is the best predictor? Why? State your reason.

In [58]:
# ANSWER

house.corr()

zip      beds     baths        sq__ft  \
zip                         1.000000 -0.066559 -0.101562 -1.015033e-01   
beds                       -0.066559  1.000000  0.653577  6.963678e-01   
baths                      -0.101562  0.653577  1.000000  7.240098e-01   
sq__ft                     -0.101503  0.696368  0.724010  1.000000e+00   
price                      -0.219807  0.440803  0.500008  6.923043e-01   
latitude                   -0.131302 -0.111139 -0.082901 -1.027984e-01   
longitude                  -0.561654  0.053003  0.106334  1.349291e-01   
actual_minus_average_price -0.047382  0.090064  0.167118  3.591270e-01   
predicted_house_prices     -0.101503  0.696368  0.724010  1.000000e+00   
residuals                  -0.207227 -0.057227 -0.001701  3.818377e-16   

                               price  latitude  longitude  \
zip                        -0.219807 -0.131302  -0.561654   
beds                        0.440803 -0.111139   0.053003   
baths                       0.500008 -0.082901   0.106334   
sq__ft                      0.692304 -0.102798   0.134929   
price                       1.000000  0.060049   0.277623   
latitude                    0.060049  1.000000   0.343975   
longitude                   0.277623  0.343975   1.000000   
actual_minus_average_price  0.529938  0.059697   0.151425   
predicted_house_prices      0.692304 -0.102798   0.134929   
residuals                   0.721606  0.181840   0.255279   

                            actual_minus_average_price  \
zip                                          -0.047382   
beds                                          0.090064   
baths                                         0.167118   
sq__ft                                        0.359127   
price                                         0.529938   
latitude                                      0.059697   
longitude                                     0.151425   
actual_minus_average_price                    1.000000   
predicted_house_prices                        0.359127   
residuals                                     0.389843   

                            predicted_house_prices     residuals  
zip                                  -1.015033e-01 -2.072269e-01  
beds                                  6.963678e-01 -5.722686e-02  
baths                                 7.240098e-01 -1.700851e-03  
sq__ft                                1.000000e+00  3.818377e-16  
price                                 6.923043e-01  7.216056e-01  
latitude                             -1.027984e-01  1.818402e-01  
longitude                             1.349291e-01  2.552786e-01  
actual_minus_average_price            3.591270e-01  3.898433e-01  
predicted_house_prices                1.000000e+00  4.401658e-16  
residuals                             4.401658e-16  1.000000e+00

#### 2.4 Find Coefficient and Intercept using Linear Regression

Using LinearRegression of sklearn.linear_model package find Coefficient and Intercept.

Create an instance of LinearRegression.

Explore following methods:

- fit
- predict
- score

In [5]:
from sklearn.linear_model import LinearRegression

##### 2.4.1 Fit predictor and target variables using linear regression 

In [59]:
# ANSWER

X = house['sq__ft'].values.reshape(-1,1)
Y = house['price']

##### 2.4.2 Find R Score

Find $R^2$ Using ```score``` method of Linear Regression.

In [60]:
# ANSWER

reg = LinearRegression().fit(X, Y)
reg.score(X, Y)

0.4792852861969368

### 3. Splitting Data

In [61]:
from sklearn.model_selection import train_test_split

#### 3.1 Create training and testing subsets

Hint: Using `train_test_split` Library.

```
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
```

In [74]:
# ANSWER
## Create training and testing subsets

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.33, random_state=42)

In [75]:
X_train

array([[1428],
       [1170],
       [1828],
       [ 588],
       [1419],
       [1991],
       [2235],
       [1130],
       [1788],
       [ 970],
       [1810],
       [1183],
       [ 779],
       [1103],
       [1315],
       [1381],
       [1232],
       [3357],
       [1354],
       [2347],
       [ 911],
       [1266],
       [1120],
       [ 881],
       [1051],
       [1440],
       [1628],
       [1229],
       [1590],
       [1416],
       [1132],
       [3009],
       [2800],
       [1150],
       [1326],
       [2960],
       [2508],
       [1559],
       [1488],
       [1859],
       [1517],
       [1064],
       [1292],
       [1289],
       [1159],
       [1043],
       [1056],
       [1110],
       [1058],
       [1520],
       [3863],
       [1202],
       [1240],
       [1200],
       [ 994],
       [1328],
       [1202],
       [2205],
       [1451],
       [1416],
       [1134],
       [ 852],
       [1245],
       [1981],
       [1140],
       [ 756],
       [27

In [76]:
X_test

array([[4400],
       [ 932],
       [1291],
       [1139],
       [1669],
       [1627],
       [1740],
       [2169],
       [2442],
       [2306],
       [1586],
       [1176],
       [1650],
       [1477],
       [1735],
       [2724],
       [1457],
       [1126],
       [2212],
       [1483],
       [3615],
       [1362],
       [1162],
       [1039],
       [ 760],
       [1039],
       [1187],
       [1393],
       [1289],
       [1616],
       [1082],
       [1144],
       [1410],
       [1401],
       [1206],
       [1601],
       [ 990],
       [1375],
       [1479],
       [1156],
       [1280],
       [1118],
       [1329],
       [2052],
       [1917],
       [1953],
       [1711],
       [ 904],
       [1250],
       [1127],
       [1209],
       [1306],
       [1637],
       [1795],
       [ 888],
       [2056],
       [2652],
       [2163],
       [1329],
       [1174],
       [ 998],
       [2367],
       [2725],
       [1264],
       [1358],
       [ 888],
       [11

In [77]:
y_train

762    188700
544    176250
615    250000
18     120000
588    220000
        ...  
71     200000
106    285000
270    370000
435    295000
102    275086
Name: price, Length: 541, dtype: int64

In [78]:
y_test

697    884790
296     65000
227    236250
336    133105
537    168750
        ...  
51     173000
334    127500
741    157500
264    339000
361    180000
Name: price, Length: 267, dtype: int64

#### 3.2 Check Shape, Sample of Test Train Data

In [79]:
# ANSWER
## Check training/test data

print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(541, 1)
(541,)
(267, 1)
(267,)


#### 3.3 Using Linear Regression Find The Score

1. Fit model using X_train, y_train
2. Find score using X_test, y_test

In [85]:
# ANSWER

reg = LinearRegression().fit(X_train, y_train)
reg.score(X_train, y_train)

0.4713873668902573

In [86]:
# ANSWER

reg.score(X_test, y_test)

0.48808261753470117

### 4. Predict House Price

Let's assume we have information of following house:

- street:	1140 EDMONTON DR
- city:	SACRAMENTO
- zip:	95833
- state:	CA
- beds:	3
- baths:	2
- sq__ft:	1204
- type:	Residential

**Predict the price of this house using linear regression model.**

In [82]:
# ANSWER

reg.predict([[1204,]])

array([182495.035267])

In [67]:
reg.coef_

array([129.13348618])

In [68]:
reg.intercept_

27221.196073751926

In [83]:
reg.coef_*1204 + reg.intercept_

array([182495.035267])

#### Find the error (Mean Squared Error)

In [84]:
# Predict price from X_test

from sklearn.metrics import mean_squared_error

mean_squared_error(X_test, y_test)


68086847562.52809



---



---



> > > > > > > > > © 2021 Institute of Data


---



---



